### ISSUE TITLE
Refactor rax module

### ISSUE THREAD

**Issue by sivel** [ansible/ansible#4577]
Just to hold the place...I'm working on this currently, but wanted to get this out there so people know it is happening.  I'd assign it to myself, but it doesn't look like I can do so.

Things that should change in the `rax` module:
1. Deprecate the `service` attribute
2. Provide more information about the instance; pyrax provides a lot, the rax module exposes few
3. Handle `active` vs `present`, if we have multiple options, use them.  Although `present` may not really make much sense, since we probably want "active" instances
4. Allow image `name` or `human_id` in addition to the image `id`.  `human_id` is preferred
5. Allow a `count` attribute, to spin up more than a single instance at a time
6. Make `active` and `present` idempotent, (See #4033 for ec2 idempotence)
7. Allow easy specification of the `group` (hostgroup) meta key used by the rax inventory script (this will also be used for the above idempotence functionality)
8. Maybe add option to only wait for network information, in some cases this may be enough for you to continue (this could be "dangerous", due to how IP assignments are retried based on assignment failures)
9. Use the `search_opts` argument for `pyrax.cloudservers.servers.list()` to perform filtering via the API
10. Provide an `instance_ids` attribute to be used for `deleted`/`absent` states (the existing functionality should remain too)

There are probably more and will be more improvements, but this should suffice for now.

**Comment by omgjlk** [comment]
present is currently idempotent, but it's based on a lot of specific details of the instance, which gets odd when you throw count into the mix.

Waiting on IP does seem dangerous.

Everything else sounds awesome! What, if anything needs to be done to operate more like nova_compute ?

**Comment by claco** [comment]
Just something for the refactor I noticed this morning. If I spin up an instance without meta: group:, then I just add the meta: group: (but we have the same name, other settings, etc), it spins up a second instance. That was unexpected.

**Comment by omgjlk** [comment]
Yeah, that's the odd bit of idempotence I went with. The group is different, so the instance is different, thus spin up a new one. It's probably not the best, but it was to match an internal tool we had.

I'm fine with other ways of achieving idempotence.

**Comment by mpdehaan** [comment]
So with regard to count, here's where we are going with the other cloud modules.

We want to make it able to not just spin up "count=N", but also be able to set the count to N.

This probably requires a flag or setting to control how count behaves, but seems optimal for treating cloud like cloud.

I'd propose exact_count=yes|no

?

Just throwing that out there.

**Comment by mpdehaan** [comment]
I also noticed there's seemingly no groups based on anything like tag.

Are tags exposable in the API?   Would be nice to have similar capabilities for grouping.

**Comment by mpdehaan** [comment]
Random thoughts, some aspects my be _wrong_ and I'm willing to have them discounted as you know more about these things :)

I would also like it if the inventory script could return flags on whether the instances were running or not , if not already. 

But as I understand it reboot is exposed and not shutdown via the RAX API?  That's hearsay though, so perhaps not correct.

"Allow easy specification of the group (hostgroup) meta key used by the rax inventory script (this will also be used for the above idempotence functionality)"

Basically like a tag?

I don't like the idea of having arbitrary idempotence attributes in general -- it's confusing to users who have to know what variables to pick and why they would choose either.

Really what we need most is a RAX chapter to match the AWS one (which is weak, and needs to be improved anyway) that shows how to put all this together, and that can ... in early prototypical form, probably make it more clear how everything is going to work together and what use cases are?

It's possible for me to accept code for a RAX chapter and not include it in the index until we are ready.  It just generates a warning on our end is all.

Another thought -- The AWS module does something a bit screwy where each instance ID is a group of one.  It would be nice to avoid this and just make it a variable on the instance.  AKA I wouldn't follow point on that one.

BTW, huge thanks for doing this.

**Comment by sivel** [comment]
In OpenStack/Rackspace they are called 'meta', and provide key/value pairs that are stored with the instance.  This is I believe what AWS calls tags.

We already look for a 'group' meta key with the rax.py inventory script, I will also use this key for grouping capability.

I'll look into `exact_count`, I was looking at the ec2 code for doing this and it appeared it is just doing this regardless, but I like an `exact_count` attribute.  I'll look into all these suggestions and work them into the code base, as well as the RAX chapter.  Although I may have to wait on the chapter, as I am not positive exactly where this is going as of now, but that should solidify over the next few weeks.

**Comment by claco** [comment]
I'm waaaay to fresh in Ansible, so this may be a silly request. But can the output of the rax module match all the things inventory can return? The details returned now on a created instance are pretty slim.

**Comment by sivel** [comment]
@claco, yes.  That is (2) on my initial list, but I will check on inventory/module info parity.

**Comment by claco** [comment]
@sivel :+1: \* 100

**Comment by angstwad** [comment]
@mpdehaan 

> I would also like it if the inventory script could return flags on whether the instances were running or not , if not already.
> 
> But as I understand it reboot is exposed and not shutdown via the RAX API? That's hearsay though, so perhaps not correct.

So, shutdown is not exposed via the API, but instances can be shutdown from the OS; the status remains available via the API.

On the topic of my pull request #4326, this status is exposed as a field.  So for this I shut down my instance from within the OS, which OpenStack does pick up and report.  Every available field from within the API, @j2sol had prepended with rax_ and embedded.  I didn't mess with this at all as it works well.

```
$ python rax.py --host ubuntu | grep rax_status
    "rax_status": "SHUTOFF",
```

**Comment by claco** [comment]
@sivel And add support for cloud_networks so I can create isolated networks. :-)

**Comment by claco** [comment]
@sivel "Deprecate the service attribute"

Are you splitting the services out into separate modules?

**Comment by Apsu** [comment]
I for one vote for separation. Less coupling, more granularity, more composability!

**Comment by claco** [comment]
Working on a rax_network right now, because I need one... right now. :-)

removing the service so hopefully it can be somewhat reused.

**Comment by sivel** [comment]
Yes, it was decided that an ansible module should only model one service. The rax module should only be used for cloud servers, each additional service should get rax_whatever.

There is another pull request somewhere to add network assignment into the rax module, but creation of networks needs to live elsewhere.

**Comment by claco** [comment]
@sivel First pass. It works. Feedback welcome. PR at some point when it has proper docs and stuff:  

https://gist.github.com/claco/a1b055a128e23f74c23e

I'll most likely be tweaking rax into a local rax_server as well to work in networks support for what rax_network is creating.

**Comment by claco** [comment]
@sivel FYI, and maybe you already know this. While hacking in networks: support to a copy of the rax module, I went down a rabbit hole so I'm sharing.

http://docs.rackspace.com/servers/api/v2/cs-devguide/content/CreateServers.html

I was expecting to take that networks structure from a playbook and pass it though the server create. And it took many tries to find the correct combo: While our API docs say uuid, it's really net-id, and the param is nics= instead of networks=  :8ball: 

In other words

```
create(name, image, flavor, nics=[{'net-id': '00000000-.....'}]
```

I'm sure this is compatery w/ v1, v2, etc. :-/

**Comment by claco** [comment]
And shortly, I'll have a rax_lb that I needed today to create cloud load balancers. Which leans me to the next q: naming.  rax_clb does not create/delete load balancers, only adds/removes nodes from it.

So, should rax_*lb do the create/delete, then updates nodes accordingly using a nodes attribute? Just looking for the vision and how these will be separated.

**Comment by sivel** [comment]
@claco That is something I have been wanting to look into.  I think that `rax_clb` should be able to handle both situations.  There shouldn't be multiple modules for dealing with clbs, just one that can handle the needs.

**Comment by sivel** [comment]
So that it is visible, the current state of the work can be seen in https://github.com/sivel/ansible/tree/rax-improvements

I'll probably squash the commits and do the pull request using a different branch, but for now it is there for viewing.

Also, as a running tally of what is changing, and example usage: https://gist.github.com/sivel/7076710

**Comment by claco** [comment]
@sivel "I think that rax_clb should be able to handle both situations. There shouldn't be multiple modules for dealing with clbs, just one that can handle the needs."

I'm actually starting to feel the opposite. I threw together a quick rax_lb module to create load balancers for two reasons: 1) feels wrong to do it in rax w/ service, and 2) there are sooooo many options that can be set during creation that adding host management into it might be a little crazy. Of course, it may just mean we need to write these modules is a much less scripty way.

There's also some nuance that needs to be split out. Should "present" always just ensure that it's create, or should it update all of the attributes of an LB? Should there be an "updated" status? If I change protocol but not anything unique (like name), should it be updated during the "present" state?

**Comment by sivel** [comment]
@claco if that is the case then it may make sense to rename the current module. It is new to 1.4 and was originally named rax_clb_nodes. I may have inadvertently made a mistake of assuming the functionalities could be combined into a single module.

Since you are closer to it than I am at this exact point, do what feels best. 

Although I recommend that you open a new issue so we can keep this issue a little more specific to the current rax module (cloud servers)

**Comment by sivel** [comment]
I think I am getting pretty close here.  I created a bunch of test plays/playbooks yesterday to verify functionality and try to catch bugs.

All of the examples are up on the gist: https://gist.github.com/sivel/7076710

**Comment by claco** [comment]
Cool!  I’m working on a PR this morning for the module to create/delete load balancers (rax_clb) and renaming rax_clb to rax_clb_nodes. I’m sure it will need many passes/reviews, but I want to at least get it rolling.

-=Chris

**Comment by sivel** [comment]
We have gotten a fair amount of testing accomplished on the refactor, and it has been validated that the refactor has not broken any previous functionality of the module.

I'll be submitting the pull request shortly.

**Comment by sivel** [comment]
Pull request submitted at #4919
